module/os/linux/zfs/zfs_vfsops.c

```diff
@@ -1869,6 +1872,9 @@ zfsvfs_teardown(zfsvfs_t *zfsvfs, boolean_t unmounting)
 		txg_wait_synced(dmu_objset_pool(zfsvfs->z_os), 0);
 	}
 	dmu_objset_evict_dbufs(zfsvfs->z_os);
+	dsl_dir_t *dd = os->os_dsl_dataset->ds_dir;
+	dd->dd_activity_cancelled = B_TRUE;
+	cv_broadcast(&dd->dd_activity_cv);
```

Comment:
We may need to do something similar if the dataset is destroyed, depending on how you address the condition where we try to wait when it's unmounted.  In any case we need to ensure that we don't have any threads waiting (with a long-hold) when we try to destroy the dataset.

There are a few more operations that also check if there are long holds (e.g. promote).  Like destroy, they can only work when the fs is not mounted, so if we ensure that we can only be waiting while it's mounted, it should address those too.

However, live receive/rollback (where the fs remains mounted) need to be considered separately.  dsl_dataset_handoff_check() will fail if there are long-holds other than the expected one (from being mounted by the ZPL, or the equivalent for zvols).  I guess we could stop waiting (like when unmounting) if we do a live-receive/rollback.